numpy/_array_api/_creation_functions.py

```diff
@@ -0,0 +1,195 @@
+    """
+    Array API compatible wrapper for :py:func:`np.asarray <numpy.asarray>`.
+
+    See its docstring for more information.
+    """
+    # _array_object imports in this file are inside the functions to avoid
+    # circular imports
+    from ._array_object import ndarray
+    if device is not None:
+        # Note: Device support is not yet implemented on ndarray
+        raise NotImplementedError("Device support is not yet implemented")
+    if copy is not None:
+        # Note: copy is not yet implemented in np.asarray
+        raise NotImplementedError("The copy keyword argument to asarray is not yet implemented")
+    if isinstance(obj, ndarray) and (dtype is None or obj.dtype == dtype):
+        return obj
+    if dtype is None and isinstance(obj, int) and (obj > 2**64 or obj < -2**63):
+        # Give a better error message in this case. NumPy would convert this
+        # to an object array.
+        raise OverflowError("Integer out of bounds for array dtypes")
+    res = np.asarray(obj, dtype=dtype)
+    if res.dtype not in _all_dtypes:
+        raise TypeError(f"The array_api namespace does not support the dtype '{res.dtype}'")
+    return ndarray._new(res)
+
+def arange(start: Union[int, float], /, stop: Optional[Union[int, float]] = None, step: Union[int, float] = 1, *, dtype: Optional[dtype] = None, device: Optional[device] = None) -> array:
+    """
+    Array API compatible wrapper for :py:func:`np.arange <numpy.arange>`.
+
+    See its docstring for more information.
+    """
+    from ._array_object import ndarray
+    if device is not None:
+        # Note: Device support is not yet implemented on ndarray
+        raise NotImplementedError("Device support is not yet implemented")
+    return ndarray._new(np.arange(start, stop=stop, step=step, dtype=dtype))
+
+def empty(shape: Union[int, Tuple[int, ...]], *, dtype: Optional[dtype] = None, device: Optional[device] = None) -> array:
+    """
+    Array API compatible wrapper for :py:func:`np.empty <numpy.empty>`.
+
+    See its docstring for more information.
+    """
+    from ._array_object import ndarray
+    if device is not None:
+        # Note: Device support is not yet implemented on ndarray
+        raise NotImplementedError("Device support is not yet implemented")
+    return ndarray._new(np.empty(shape, dtype=dtype))
+
+def empty_like(x: array, /, *, dtype: Optional[dtype] = None, device: Optional[device] = None) -> array:
+    """
+    Array API compatible wrapper for :py:func:`np.empty_like <numpy.empty_like>`.
+
+    See its docstring for more information.
+    """
+    from ._array_object import ndarray
+    if device is not None:
+        # Note: Device support is not yet implemented on ndarray
+        raise NotImplementedError("Device support is not yet implemented")
+    return ndarray._new(np.empty_like(x._array, dtype=dtype))
```

Comment:
@rgommers's opinion was that we shouldn't have instance checks (we do have type signatures, and the `_array` at least means it will fail loudly with AttributeError). I personally don't feel strongly either way.

Reply:
I'm struggling to remember the context, was this just about runtime cost of doing a check, or something else?